### PR TITLE
Merge "DNSLink" and "Redirect gateway-like resouces" settings

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/31097

2 settings are merged into a new one, so if any of previous settings were enabled, we switch the new setting on

Sec review : https://github.com/brave/security/issues/1315

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1) Take fresh browser profile
2) Enable "Automatically redirect to IPFS pages via DNSLink when possible" setting
3) Update browser
4) Check that there is a new setting which is DISABLED 
5) Repeat with "Redirect IPFS resources to the configured IPFS gateway" settings

..
6) Enable both DNSLink and "Redirect IPFS resources" and check that new setting is enabled

### CODE REVIEW

File: app/brave_settings_strings.grdp [hunk 1]
```diff
@@ -1190,19 +1190,13 @@
   <message name="IDS_SETTINGS_IPFS_AUTO_FALLBACK_TO_GATEWAY_DESC" desc="The description for the switch of automatically fallback to IPFS public gateway">
     Automatically fall back to public IPFS gateway if your Brave local IPFS node cannot be reached.
   </message>
-  <message name="IDS_SETTINGS_IPFS_AUTO_REDIRECT_GATEWAY_LABEL" desc="The text label for redirecting IPFS resources to the configured gateway">
-    Redirect IPFS resources to the configured IPFS gateway
+  <message name="IDS_SETTINGS_IPFS_AUTO_REDIRECT_TO_GATEWAY_WHEN_POSSIBLE_LABEL">
+    Automatically redirect resources available on IPFS network to the configured gateway.
```
Comment (mkarolin) [inline app/brave_settings_strings.grdp:1194]: Suggestion: `Automatically redirect requests for IPFS network resources to the configured gateway.` cc: @rmcfadden3

File: browser/resources/settings/brave_ipfs_page/brave_ipfs_page.html [hunk 2]
```diff
@@ -64,18 +64,11 @@
...
-      pref="{{prefs.brave.ipfs.auto_redirect_gateway}}"
-      label="$i18n{ipfsAutoRedirectGatewayLabel}"
-      sub-label="$i18n{ipfsAutoRedirectGatewayDesc}">
-    </settings-toggle-button>
-    <settings-toggle-button id="ipfsAutoRedirectDNSLink"
-      class="cr-row"
-      pref="{{prefs.brave.ipfs.auto_redirect_dnslink}}"
-      label="$i18n{ipfsAutoRedirectDNSLinkLabel}"
-      sub-label="$i18n{ipfsAutoRedirectDNSLinkDesc}"
-      learn-more-url="$i18n{ipfsDNSLinkLearnMoreURL}">
```
Comment (diracdeltas) [inline browser/resources/settings/brave_ipfs_page/brave_ipfs_page.html:78]: should we keep the "learn more" url?

File: components/ipfs/ipfs_service.cc [hunk 3]
```diff
@@ -159,15 +159,28 @@ IpfsService::~IpfsService() {
...
 
+// static
+void IpfsService::MigrateProfilePrefs(PrefService* pref_service) {
+  if (pref_service->HasPrefPath(kIPFSAutoRedirectToConfiguredGateway)) {
+    return;
+  }
+  pref_service->SetBoolean(
+      kIPFSAutoRedirectToConfiguredGateway,
+      pref_service->GetBoolean(kIPFSAutoRedirectDNSLink) ||
+          pref_service->GetBoolean(kIPFSAutoRedirectGateway));
```
Comment (diracdeltas) [inline components/ipfs/ipfs_service.cc:170]: i think this should prompt if only one of them was on, see comment here: https://github.com/brave/security/issues/1315#issuecomment-1595239760

File: browser/ipfs/test/ipfs_service_browsertest.cc [hunk 4]
```diff
@@ -1666,4 +1676,24 @@ IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest, DNSResolversConfig) {
   }
 }
 
+IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest, MigrateProfilePrefs) {
+  auto* prefs = browser()->profile()->GetPrefs();
```
Comment (fallaciousreasoning) [inline browser/ipfs/test/ipfs_service_browsertest.cc:1680]: Might be worth checking the `false` `false` and `true` `true` cases too, just to make sure we've covered everything.

File: browser/resources/settings/brave_ipfs_page/brave_ipfs_page.html (hunk 2 above)
Comment (fallaciousreasoning) [reply browser/resources/settings/brave_ipfs_page/brave_ipfs_page.html:78]: And if not, do we need to remove something from the help pages?

**Review by fallaciousreasoning — APPROVED** [review]
LGTM :smile:

**Review by diracdeltas — APPROVED** [review]

**Pull request merged**